Implement MBC3 (more or less) and MBC5

LukasKalbertodt opened the pull request:
MBC3 has a real time clock. But it's not yet clear how we would measure how much time has passed. So I'd say we leave it as a TODO for now.

LukasKalbertodt committed:
Respect `ram_enabled` bit for MBC1

LukasKalbertodt committed:
Add MBC5

LukasKalbertodt committed:
Fix minor bug in PPU (forgot to reset a value)

LukasKalbertodt committed:
Add MBC3 (with dummy RTC data)

The real time clock was not properly implemented and always returned
date 0 basically. Implementing this is not super easy for all platforms,
as WASM can't easily get a current time. So we have to discuss how
to proceed on this one.

jovobe commented on core/src/mbc/mbc3.rs:
```diff
@@ -0,0 +1,263 @@
+    pub(crate) fn new(data: &[u8], rom_size: RomSize, ram_size: RamSize) -> Self {
+        assert!(rom_size <= RomSize::Banks128, "More than 128 banks, but only MBC3!");
+        assert!(
+            rom_size.len() == data.len(),
```
Does this assertion make sense? In my opinion the `data` needs to fit in the `rom_size` but the size does not have to match exactly.

LukasKalbertodt replied:
I wasn't sure about that either. I decided to add the stricter assert to see what happens. It seems that ROM files always have exactly the correct size, even if part of that data is not used (probably just 0s). We can of course relax this assertion, but I'm kind of curious to see if there are ROMs where this does not match.

jovobe replied:
That's interesting! In this case I like your implementation - let's leave it this way.

jovobe commented on core/src/mbc/mbc3.rs:
```diff
@@ -0,0 +1,263 @@
+                let idx = self.ram_bank as usize * 0x2000 + addr.get() as usize;
+                if idx < self.ram.len() {
+                    self.ram[idx] = byte;
+                }
```
I think we should `warn!` here (in `else` case) for debug purposes.

jovobe commented on core/src/mbc/mbc5.rs:
```diff
@@ -0,0 +1,121 @@
+impl Mbc5 {
+    pub(crate) fn new(data: &[u8], rom_size: RomSize, ram_size: RamSize) -> Self {
+        assert!(
+            rom_size.len() == data.len(),
```
Same as on `Mbc3`.

jovobe commented on core/src/mbc/mbc5.rs:
```diff
@@ -0,0 +1,121 @@
+            rom_size.len() == data.len(),
+            "Length of cartridge doesn't match length specified in ROM size header",
+        );
+
```
Why don't you check for `rom_size <= RomSize::???` here?

LukasKalbertodt replied:
Because MBC5 supports up to 512 banks which is the largest value. However, I added the check now anyway in case we extend the `RomSize` enum at some point.

jovobe commented on core/src/mbc/mbc5.rs:
```diff
@@ -0,0 +1,121 @@
+        let idx = self.ram_bank as usize * 0x2000 + addr.get() as usize;
+        if idx < self.ram.len() {
+            self.ram[idx] = byte;
+        }
```
Same as on `Mbc3`.

jovobe commented on core/src/mbc/mbc5.rs:
```diff
@@ -0,0 +1,121 @@
+        );
+
+        let rom: Vec<_> = data.iter().cloned().map(Byte::new).collect();
+        let ram = vec![Byte::zero(); ram_size.len()];
```
Here should be a check for the RAM size. Allowed sizes are `64kbit, 256kbit and 1mbit.` according to [[1]](http://gbdev.gg8.se/wiki/articles/MBC5). And I think our `RamSize` is currently limited to `RamSize::Kb32` this should be changed. When this is changed `Mbc1` and `Mbc3` should be limited to a ram size of `RamSize::Kb32` according to [[2]](http://gbdev.gg8.se/wiki/articles/MBC1) and [[3]](http://gbdev.gg8.se/wiki/articles/MBC3).

- \[1\]: http://gbdev.gg8.se/wiki/articles/MBC5
- \[2\]: http://gbdev.gg8.se/wiki/articles/MBC1
- \[3\]: http://gbdev.gg8.se/wiki/articles/MBC3

LukasKalbertodt committed:
Extend `RamSize` to include `Kb64` and `Kb128`

LukasKalbertodt committed:
Add RAM size checks for MBC1, 3 and 5

LukasKalbertodt committed:
Add warnings when games write to illegal RAM areas

LukasKalbertodt committed:
Add ROM size check to MBC5

LukasKalbertodt committed:
Remove now stabilized `edition` Cargo feature gate

This requires a new nightly compiler. It works with rustc 1.30.0-nightly
(f2302daef 2018-09-12), potentially also slightly older versions.

LukasKalbertodt commented:
FYI: my last commit removes the `cargo-features` from all `Cargo.toml`s. This means that a newer nightly version is required to compile this. I use `rustc 1.30.0-nightly (f2302daef 2018-09-12)`, but it might also work with slightly older versions.

jovobe commented on core/src/cartridge.rs:
```diff
@@ -188,13 +188,26 @@ impl PartialOrd for RomSize {
+    Kb32,
+
+    /// 16 banks.
+    Kb128,
```
According to [\[1\]](http://gbdev.gg8.se/wiki/articles/MBC5) this size does not exist. This should be replaced by 256Kb and 1Mb.

- \[1\]: http://gbdev.gg8.se/wiki/articles/MBC5

LukasKalbertodt replied:
The sizes on that page are given in K**bit** and M**bit**. Divide by 8 to get 32 KiB and 128 KiB.

jovobe replied:
My fault, I interpreted your comment on the enum wrong `KiB` means of course Byte 🤦‍♂️

jovobe commented on core/src/mbc/mbc5.rs:
```diff
@@ -27,10 +28,16 @@ pub(crate) struct Mbc5 {
             "Length of cartridge doesn't match length specified in ROM size header",
         );
+        assert!(
+            [RamSize::None, RamSize::Kb8, RamSize::Kb32, RamSize::Kb128].contains(&ram_size),
```
Wouldn't it be better to whitelist the allowed sizes here? In this case we won't have to update the assertion every time the enum changes.

LukasKalbertodt replied:
This is a whitelist. `assert!` makes sure the boolean expression is `true`. So we make sure that `ram_size` is one of the listed ones. But yeah, I'm confused by asserts every time again, mostly because the messages of the assert talks about the failure case, that is: when the boolean expression is `false`.

jovobe replied:
Of course 🙈
